lib/temporal/client.rb

```diff
@@ -0,0 +1,382 @@
+require 'temporal/interceptor/chain'
+require 'temporal/workflow/handle'
+require 'temporal/workflow/execution_info'
+require 'temporal/workflow/id_reuse_policy'
+require 'temporal/workflow/query_reject_condition'
+
+module Temporal
+  class Client
+    attr_reader :namespace, :identity
+
+    # TODO: More argument to follow for converters, codecs, etc
+    def initialize(connection, namespace, interceptors: [])
+      @connection = connection
+      @namespace = namespace
+      @identity = "#{Process.pid}@#{Socket.gethostname}"
+      @interceptor_chain = Interceptor::Chain.new(interceptors)
+      @converter = Converter.new
+    end
+
+    def start_workflow(
+      workflow,
+      *args,
+      id:,
+      task_queue:,
+      execution_timeout: nil,
+      run_timeout: nil,
+      task_timeout: nil,
+      id_reuse_policy: Workflow::IDReusePolicy::ALLOW_DUPLICATE,
+      retry_policy: nil,
+      cron_schedule: '',
+      memo: nil,
+      search_attributes: nil,
+      headers: {},
+      start_signal: nil,
+      start_signal_args: []
+    )
+      retry_policy&.validate!
+
+      input = Interceptor::Client::StartWorkflowInput.new(
+        workflow: workflow,
+        args: args,
+        id: id,
+        task_queue: task_queue,
+        execution_timeout: execution_timeout,
+        run_timeout: run_timeout,
+        task_timeout: task_timeout,
+        id_reuse_policy: id_reuse_policy,
+        retry_policy: retry_policy,
+        cron_schedule: cron_schedule,
+        memo: memo,
+        search_attributes: search_attributes,
+        headers: headers,
+        start_signal: start_signal,
+        start_signal_args: start_signal_args,
+      )
+
+      interceptor_chain.invoke(:start_workflow, input) do |i|
+        handle_start_workflow(i)
+      end
+    end
+
+    def describe_workflow(id, run_id)
+      input = Interceptor::Client::DescribeWorkflowInput.new(
+        id: id,
+        run_id: run_id,
+      )
+
+      interceptor_chain.invoke(:describe_workflow, input) do |i|
+        handle_describe_workflow(i)
+      end
+    end
+
+    def query_workflow(
+      id,
+      run_id,
+      query:,
+      args: [],
+      reject_condition: Workflow::QueryRejectCondition::NONE
+    )
+      input = Interceptor::Client::QueryWorkflowInput.new(
+        id: id,
+        run_id: run_id,
+        query: query,
+        args: args,
+        reject_condition: reject_condition,
+      )
+
+      interceptor_chain.invoke(:query_workflow, input) do |i|
+        handle_query_workflow(i)
+      end
+    end
+
+    def signal_workflow(id, run_id, signal:, args: [])
+      input = Interceptor::Client::SignalWorkflowInput.new(
+        id: id,
+        run_id: run_id,
+        signal: signal,
+        args: args,
+      )
+
+      interceptor_chain.invoke(:signal_workflow, input) do |i|
+        handle_signal_workflow(i)
+      end
+    end
+
+    def cancel_workflow(id, run_id: nil, first_execution_run_id: nil, reason: nil)
```

Comment:
Basically using keyword arguments for optional arguments here, since you should be able to just `client.cancel_workflow(workflow_id)`. Basically separating main arguments and options (but I admit it's not always consistent)